### PR TITLE
	modified:   plots.py

### PR DESCRIPTION
**Pull request by DeliciousHair**
```
Expanded plots.autocorrplot to accept multi-dimension parameters (up to 10 for now)
```

### CODE REVIEW

**Comment by twiecki** [comment]
Thanks @DeliciousHair. Can you paste a screenshot of what this looks like?

**Comment by DeliciousHair** [comment]
Sure, but excuse the really crummy demo data. Let's say I have:

``` python
In [85]: tr.varnames
Out[85]: ['mu_interval', 'sd_interval', 'mu', 'sd', 'delta_mu']

In [86]: tr['mu']
Out[86]: 
array([[ 62.04236681,  57.08562924],
       [ 62.04236681,  57.08562924],
       [ 62.59687376,  59.38829821],
       ..., 
       [ 62.67018591,  58.52952785],
       [ 62.67018591,  58.52952785],
       [ 62.67018591,  58.52952785]])
```

this modification will now generate the autocorrelation plot for both `mu[:, 0]` and `mu[:, 1]` (or everything for that matter)

``` python
In [90]: autocorrplot(tr, vars=['mu', 'delta_mu'], burn=5000, max_lag=5)
Out[90]: 
(<matplotlib.figure.Figure at 0x7f6a911a8e50>,
 array([[<matplotlib.axes._subplots.AxesSubplot object at 0x7f6a91354550>],
        [<matplotlib.axes._subplots.AxesSubplot object at 0x7f6a9117d350>],
        [<matplotlib.axes._subplots.AxesSubplot object at 0x7f6a910de7d0>]], dtype=object))
```

autocorr_demo

Hm, I see I didn't put the variable names on as titles.  Ah well, that's a trivial change :-)

**Comment by jsalvatier** [comment]
Cool! Two comments: 
1. There seems to be a build error: https://travis-ci.org/pymc-devs/pymc3/jobs/76077291

```
ERROR: pymc3.tests.test_examples.test_examples0('pymc3.examples.GHME_2013',)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/testenv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/pymc-devs/pymc3/pymc3/tests/test_examples.py", line 15, in check_example
    example.run("short")
  File "/home/travis/build/pymc-devs/pymc3/pymc3/examples/GHME_2013.py", line 115, in run
    autocorrplot(trace, vars = [coeff_sd,sd ])
  File "/home/travis/build/pymc-devs/pymc3/pymc3/plots.py", line 189, in autocorrplot
    for var in vars] for item in sub]
  File "/home/travis/build/pymc-devs/pymc3/pymc3/plots.py", line 177, in _handle_array_varnames
    if trace[0][val].__class__ is np.ndarray:
KeyError: coeff_sd
```
1. I've seen other packages do this too, but is there a good reason to have both the positive and negative lag correlations? Can't we just do the positive ones? Maybe that's a separate change since it seems like you didn't make that.

**Comment by DeliciousHair** [comment]
1. Ah yes, that makes sense--I couldn't figure out what the seemingly redundant line

``` python
    else:
        vars = [str(var) for var in vars]
```

was for in the existing function.  Will fix that and the labeling now.
1. I didn't touch this part of things at all, but can have a look if you like. You're correct though, given the application taking the pos. and neg. lags should not be necessary as they would be (basically) symmetric and either direction would convey the information one is looking for anyway.

Will get to work now, guys gotta get some sleep at some point here!

**Comment by DeliciousHair** [comment]
Hrm...

I'm now getting a lot of fails due to problems with theano which I don't understand. Can somebody please have a look at this and possibly provide some insight into what I've actually broken?

**Comment by DeliciousHair** [comment]
Aurrgh!

I have reverted back to the original version I used that almost worked, and it now fails most of the tests due to the cryptic theano problems. I am at a loss, as I don't understand at all what is not working.  Any input would be greatly appreciated.

File: pymc3/plots.py [hunk 1]
```diff
@@ -168,34 +167,53 @@ def autocorrplot(trace, vars=None, max_lag=100, burn=0, ax=None):
...
+                yield val + '_{0}'.format(i)
+        else:
+            yield val
+            
     if vars is None:
-        vars = trace.varnames
+        vars = [item  for sub in [[i for i in _handle_array_varnames(var)]
+                    for var in trace.varnames] for item in sub]
     else:
-        vars = [str(var) for var in vars]
```
Comment (jsalvatier) [inline pymc3/plots.py:181]: I think its partly the same thing, you need to make sure you're dealing with strings when you think you're dealing with strings.

File: pymc3/plots.py [hunk 2]
```diff
@@ -168,34 +167,53 @@ def autocorrplot(trace, vars=None, max_lag=100, burn=0, ax=None):
...
-
     ax : matplotlib axes
-
     """
     
     import matplotlib.pyplot as plt
-
+        
+    def _handle_array_varnames(val):
+        #val = str(val)
```
Comment (jsalvatier) [inline pymc3/plots.py]: Or you can add this back.

Comment (DeliciousHair) [inline pymc3/plots.py]: This is what I'm saying, the commented out line 176 is what I put in to deal with the original, single failed test. That addition led to almost all tests failing with the theano errors I cannot figure out,  and now when I comment the line back out I still get all the theano errors. So I've "magically" gone from a small problem to a seemingly large one with no change.

The current test (that I'm now failing yet again) has all the labeling commented out, which was the only other alteration I made with, but I cannot see how something as benign as `ax.set_title()` should cause any test to fail.

I'll persist, but I really would like some insight into these theano errors if anyone has any ideas :-)

**Comment by DeliciousHair** [comment]
Ha, it's not me!

Out of pure frustration I have just reverted this function to the current one in master to see if it would pass the tests, as I cannot see how my alterations should be triggering these failures at all. Sure enough, the current build fails with the same errors I'm getting.

Thus, once this batch of failures is complete I'll push my "last" fix which should deal with the original problem correctly, and leave it for now until somebody more knowledgeable is able to figure out the current build errors that don't seem to be my fault.

**Comment by twiecki** [comment]
Just restarted the build on master and indeed seeing problems. Investigating...

**Comment by twiecki** [comment]
https://github.com/pymc-devs/pymc3/issues/803 does this match with what you're seeing?

**Comment by DeliciousHair** [comment]
Yup, that's them. Nasty! And I'm afraid I can be of very little use in helping you resolve this as I know very little about theano :-(

**Comment by twiecki** [comment]
Yeah, we'll figure it out. Sorry it caused you so much confusion.

**Comment by DeliciousHair** [comment]
No problem, I was very confused how adding labels would cause so much trouble, but assumed it was possible given my lack of knowledge about theano and more specifically how you've incorporated it into pymc3--I'm just using the tool at this point :-)

**Comment by jsalvatier** [comment]
Sorry about the trouble, Delicious. Perhaps you can just test it out locally and then we'll merge it when we resolve this?

**Comment by DeliciousHair** [comment]
No trouble at all! Just a bit frustrated that I can't fix it myself. As is, the current pr that's here now (that has failed testing) should be fine so I'll just try to restart it when this mystery theano problem is resolved. In the meantime I'll get the positive-lags-only feature working, making it a keyword argument like `symmetric_plot=False`.

Cheers!

**Comment by jsalvatier** [comment]
Sounds good!

**Comment by twiecki** [comment]
@DeliciousHair This is fixed by reverting travis to an older version. Can you rebase this branch off of current master?

**Comment by DeliciousHair** [comment]
@takluyver Yup, can do! Probably not today though as my schedule is a bit messed up at the moment but I'll find a few minutes as soon as I can. Thanks!

**Pull request closed without merging**